src/transpile/fusion.hpp

```diff
@@ -32,6 +32,645 @@ using oplist_t = std::vector<op_t>;
+        fusing_op_idxs.push_back(fusing_op_idx); // All the qubits of tgt_op are in fusing_qubits
+      else if (fusing_qubits.empty())
+          break;
+    }
+
+    std::reverse(fusing_op_idxs.begin(), fusing_op_idxs.end());
+    fusing_qubits.clear();
+    fusing_qubits.insert(fusing_qubits.end(), ops[op_idx].qubits.begin(), ops[op_idx].qubits.end());
+
+    // 3. fuse operations with forwarding
+    for (int fusing_op_idx = op_idx + 1; fusing_op_idx < fusion_end; ++fusing_op_idx) {
+      auto& tgt_op = ops[fusing_op_idx];
+      if (tgt_op.type == optype_t::nop)
+        continue;
+      if (!method.can_apply(tgt_op, max_fused_qubits))
+        break;
+      // check all the qubits are in fusing_qubits
+      if (!exclude_escaped_qubits(fusing_qubits, tgt_op))
+        fusing_op_idxs.push_back(fusing_op_idx); // All the qubits of tgt_op are in fusing_qubits
+      else if (fusing_qubits.empty())
+          break;
+    }
+
+    if (fusing_op_idxs.size() <= 1)
+      continue;
+
+    // 4. generate a fused operation
+    allocate_new_operation(ops, op_idx, fusing_op_idxs, method, false);
+
+    fused = true;
+  }
+
+  return fused;
+}
+
+class DiagonalFusion : public Fuser {
+public:
+  DiagonalFusion() = default;
+
+  virtual ~DiagonalFusion() = default;
+
+  virtual std::string name() const override { return "diagonal"; };
+
+  virtual void set_config(const json_t &config) override;
+
+  virtual bool aggregate_operations(oplist_t& ops,
+                                    const int fusion_start,
+                                    const int fusion_end,
+                                    const uint_t max_fused_qubits,
+                                    const FusionMethod& method) const override;
+
+private:
+  bool is_diagonal_op(const op_t& op) const;
+
+  int get_next_diagonal_end(const oplist_t& ops, const int from, std::set<uint_t>& fusing_qubits) const;
+
+  const std::shared_ptr<FusionMethod> method;
+  uint_t min_qubit = 3;
+  bool active = true;
+};
+
+void DiagonalFusion::set_config(const json_t &config) {
+  if (JSON::check_key("fusion_enable.diagonal", config))
+    JSON::get_value(active, "fusion_enable.diagonal", config);
+  if (JSON::check_key("fusion_min_qubit.diagonal", config))
+    JSON::get_value(min_qubit, "fusion_min_qubit.diagonal", config);
+}
+
+bool DiagonalFusion::is_diagonal_op(const op_t& op) const {
+
+  if (op.type == Operations::OpType::gate) {
+    if (op.name == "p" || op.name == "cp" || op.name == "u1" || op.name == "cu1"
+        || op.name == "mcu1" || op.name== "rz" || op.name== "rzz")
+      return true;
+    if (op.name == "u3")
+      return op.params[0] == std::complex<double>(0.) && op.params[1] == std::complex<double>(0.);
+    else
+      return false;
+  }
+
+  if (op.type == Operations::OpType::diagonal_matrix)
+    return true;
+
+  if (op.type == Operations::OpType::matrix)
+    return op.mats.size() == 1 && Utils::is_diagonal(op.mats[0], 1e-7);
```

Comment:
Won't it be expensive to check every matrix is diagonal?

Reply:
I took it.